From: Benjamin Coddington <bcodding@redhat.com>
Date: Wed, 6 Apr 2016 11:32:52 -0400
Subject: [PATCH] SUNRPC: init xdr_stream for zero iov_len, page_len
Git-commit: 06ef26a0e35199cc0445d7d5708533a12af8ff5b
Patch-mainline: v4.7
References: bsc#11303356

An xdr_buf with head[0].iov_len = 0 and page_len = 0 will cause
xdr_init_decode() to incorrectly setup the xdr_stream.  Specifically,
xdr->end is never initialized.

Signed-off-by: Benjamin Coddington <bcodding@redhat.com>
Signed-off-by: Anna Schumaker <Anna.Schumaker@Netapp.com>
Acked-by: NeilBrown <neilb@suse.com>

---
 net/sunrpc/xdr.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/net/sunrpc/xdr.c
+++ b/net/sunrpc/xdr.c
@@ -796,6 +796,8 @@ void xdr_init_decode(struct xdr_stream *
 		xdr_set_iov(xdr, buf->head, buf->len);
 	else if (buf->page_len != 0)
 		xdr_set_page_base(xdr, 0, buf->len);
+	else
+		xdr_set_iov(xdr, buf->head, buf->len);
 	if (p != NULL && p > xdr->p && xdr->end >= p) {
 		xdr->nwords -= p - xdr->p;
 		xdr->p = p;
